### PR TITLE
Source Facebook Marketing: Fix number and integer fields in schemas

### LINKED ISSUE #7533: Source Facebook Marketing 0.2.21: JobTimeout with 0 seconds and Job Failed message return
## Enviroment
- **Airbyte version**:  0.30.20-alpha
- **OS Version / Instance**: GCP
- **Deployment**: Docker
- **Source Connector and version**: Facebook Marketing 0.2.21
- **Destination Connector and version**: Snowflake
- **Severity**: Very Low / Low / Medium / **High** / Critical
- **Step where error happened**: Deploy / **Sync job** / Setup new connection / Update connector / Upgrade Airbyte

## Current Behavior
Job fails in a random fashion.

The last sync: 
image

The last week:
image

## Expected Behavior
Connector should handle error messgae from API and give me a good feedback to understand the failure reason.

## Logs
```
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 - Traceback (most recent call last):
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/airbyte/integration_code/main.py", line 13, in <module>
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     launch(source, sys.argv[1:])
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 108, in launch
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     for message in source_entrypoint.run(parsed_args):
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 99, in run
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     for message in generator:
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 108, in read
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     raise e
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 104, in read
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     internal_config=internal_config,
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 134, in _read_stream
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     for record in record_iterator:
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 206, in _read_full_refresh
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     for record in records:
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/airbyte/integration_code/source_facebook_marketing/streams.py", line 330, in read_records
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     result = self.wait_for_job(stream_slice["job"])
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     ret = target(*args, **kwargs)
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -   File "/airbyte/integration_code/source_facebook_marketing/streams.py", line 371, in wait_for_job
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     raise JobTimeoutException(f"AdReportRun {job} failed after {runtime.in_seconds()} seconds.")
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 - source_facebook_marketing.common.JobTimeoutException: AdReportRun <AdReportRun> {
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     "account_id": "255091422171865",
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     "async_percent_completion": 0,
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     "async_status": "Job Failed",
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     "date_start": "2020-10-21",
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     "date_stop": "2020-10-27",
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     "id": "374882567717385",
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     "report_run_id": "374882567717385",
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 -     "time_ref": 1635764820
2021-11-01 11:08:01 ERROR () LineGobbler(voidCall):65 - } failed after 0 seconds.
```

[logs-5534-0.txt](https://github.com/airbytehq/airbyte/files/7455511/logs-5534-0.txt)

## Steps to Reproduce
1.
2.
3.

## Are you willing to submit a PR?

Remove this with your answer.

### PR DESCRIPTION
**Pull request by htrueman**
## What
Closes #7533

## How
Update `number` and `integer` fields type.

## Recommended reading order
1. schemas files

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by sherifnada** [comment]
@htrueman can you explain how this fixes the issue?

**Comment by Zirochkaa** [comment]
`FBMarketingStream` already has a `transformer` in it. It converts fields in records to appropriate types from schema files. So if field in record has string type, for example `0.801221`, then `transformer` will convert it to `number` type (float in python). So how adding `string` type to some fields will help in fixing this issue?
btw, `transformer` spams a lot of `WARN` logs.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/schemas/ad_creatives.json [hunk 1]
```diff
@@ -85,7 +85,7 @@
         "link_data": {
           "properties": {
             "additional_image_index": {
-              "type": ["null", "integer"]
+              "type": ["null", "integer", "string"]
```
Comment (avida) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/schemas/ad_creatives.json:88]: so this is integer or string? Maybe better to add schema type transformation to cast this field? https://docs.airbyte.io/connector-development/cdk-python/schemas#type-transformation

**Comment by htrueman** [comment]
@sherifnada I see that the fields with `$ref` in schema is failing. Going to research it deeper

**Comment by htrueman** [comment]
UI issue with field types appeared. New issue added to fix that https://github.com/airbytehq/airbyte/issues/7672

**Comment by htrueman** [comment]
@sherifnada added a retry job logics. This should help in cases where it's possible. But if now then we couldn't fix that, because the issue is on the facebook side.

Working on unit_tests to cover this new job retry logics.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py [hunk 2]
```diff
@@ -367,10 +369,21 @@ def wait_for_job(self, job) -> AdReportRun:
...
+                if self.times_job_restarted.get(time_range, 0) < 6:
+                    params = deep_merge(
+                        {"time_range": {"since": job["date_start"], "until": job["date_stop"]}},
+                        self.request_params(stream_state=stream_state),
+                    )
+                    restart_job = self._create_insights_job(params)
+                    self.running_jobs.append(restart_job)
+                    self.times_job_restarted[time_range] += 1
+
+                if job["async_status"] == "Job Failed":
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:383]: should this be `elif`?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py [hunk 3]
```diff
@@ -367,10 +369,21 @@ def wait_for_job(self, job) -> AdReportRun:
...
+            if job["async_status"] in ["Job Failed", "Job Skipped"]:
+                time_range = (job["date_start"], job["date_stop"])
+                if self.times_job_restarted.get(time_range, 0) < 6:
+                    params = deep_merge(
+                        {"time_range": {"since": job["date_start"], "until": job["date_stop"]}},
+                        self.request_params(stream_state=stream_state),
+                    )
+                    restart_job = self._create_insights_job(params)
+                    self.running_jobs.append(restart_job)
+                    self.times_job_restarted[time_range] += 1
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:381]: so the idea is to have a global max of 6 retries? Not blocking for this PR but we should make this per-job

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/schemas/ad_creatives.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (htrueman) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/schemas/ad_creatives.json]: I've reverted these changed

**Pull request merged**